### PR TITLE
Validate type against StreamDescriptor in OrcRecordReader

### PR DESCRIPTION
**Pull request by geraint0923**
Validate type against StreamDescriptor in OrcRecordReader.createStreamReader.

Note that in some of the tests in presto-raptor(especially in TestRaptorIntegrationSmokeTest), they try to match OrcType.Long with Type.Date which is not allowed in this validation process, and cause the failures of these tests.

### CODE REVIEW

**Comment by geraint0923** [comment]
The most updated commit will match OrcType.Long with Type.DATE and Type.TIMESTAMP to accommodate presto-raptor, so all the tests pass now.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 1]
```diff
@@ -393,12 +398,79 @@ private void advanceToNextStripe()
...
+                break;
+            case StandardTypes.TIMESTAMP:
+                valid = orcTypeKind == OrcTypeKind.TIMESTAMP ||
+                        orcTypeKind == OrcTypeKind.LONG;
+                break;
+            case StandardTypes.VARBINARY:
+                valid = orcTypeKind == OrcTypeKind.BINARY;
+                break;
+        }
+        return valid;
```
Comment (haozhun) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Change
```
valid = xxxxxxx
break;
```
to
```
return xxx
```
and then remove this line.

Comment (haozhun) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Throw an exception if execution flow ever reaches here.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 2]
```diff
@@ -393,12 +398,79 @@ private void advanceToNextStripe()
         for (int columnId = 0; columnId < rowType.getFieldCount(); columnId++) {
             if (includedColumns.containsKey(columnId)) {
                 StreamDescriptor streamDescriptor = streamDescriptors.get(columnId);
+
+                Type type = includedColumns.get(columnId);
+                checkState(validateOrcType(streamDescriptor, type.getTypeSignature()), "OrcType and Type do not match");
```
Comment (haozhun) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]:
```
if (...) {
    throw new PrestoException(HIVE_BAD_DATA, "...");
}

File: presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java [hunk 3]
```diff
@@ -247,7 +248,8 @@ public void doIntegration(TestingOrcDataSource orcDataSource, DataSize maxMergeD
                 () -> { });
     }
 
-    private static class FakeOrcDataSource
+    @VisibleForTesting
+    static class FakeOrcDataSource
```
Comment (haozhun) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestCachingOrcDataSource.java]: Rename this to `DummyOrcDataSource`. Add a private constructor. Make this class standalone in test/`com.facebook.presto.orc`. Do this in a separate commit. But don't create a new pull request.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+public class TestOrcRecordReader
+{
+    @Test
+    public void testValidateOrcType()
+    {
+        final List<OrcType> types = ImmutableList.of(
+                // 0
+                new OrcType(OrcType.OrcTypeKind.STRUCT,
+                        ImmutableList.of(1, 2, 3, 7),
+                        ImmutableList.of(
```
Comment (haozhun) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java]: Put arguments of this function on one line. Do the same for 2 and 7 as well.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+@Test
+public class TestOrcRecordReader
+{
+    @Test
+    public void testValidateOrcType()
+    {
+        final List<OrcType> types = ImmutableList.of(
+                // 0
+                new OrcType(OrcType.OrcTypeKind.STRUCT,
```
Comment (haozhun) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java]: When you put arguments on multiple lines, always put the argument on its own line.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                        ImmutableList.of(1, 2, 3, 7),
+                        ImmutableList.of(
+                                "col_0_int",
+                                "col_1_map",
+                                "col_2_list",
+                                "col_3_struct"
+                        )),
+                // 1
+                new OrcType(OrcType.OrcTypeKind.INT,
```
Comment (haozhun) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java]: for 1/4/5/6/8/9, put argument on the same line.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertThrows;
+
+@Test
+public class TestOrcRecordReader
+{
+    @Test
+    public void testValidateOrcType()
+    {
+        final List<OrcType> types = ImmutableList.of(
```
Comment (haozhun) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java]: Remove `final`. Since recent version of Java, this is no longer necessary.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+                3, new RowType(ImmutableList.of(DoubleType.DOUBLE, BooleanType.BOOLEAN), Optional.of(ImmutableList.of("nested_col_0", "nested_col_1")))
+        );
+        final Map<Integer, Type> includedColumnsBad1 = ImmutableMap.of(
+                0, BooleanType.BOOLEAN,
+                1, new MapType(VarcharType.createVarcharType(2), DateType.DATE),
+                2, new ArrayType(DoubleType.DOUBLE)
+        );
+        final Map<Integer, Type> includedColumnsBad2 = ImmutableMap.of(
+                0, BigintType.BIGINT,
+                1, new ArrayType(DoubleType.DOUBLE),
```
Comment (haozhun) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java]: `1, new ArrayType(DoubleType.DOUBLE), // should be map<varchar,date>`
Same for the other bad ones.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        final Map<Integer, Type> includedColumnsBad5 = ImmutableMap.of(
+                0, BigintType.BIGINT,
+                1, new MapType(VarcharType.createVarcharType(2), DateType.DATE),
+                2, new ArrayType(DoubleType.DOUBLE),
+                3, new RowType(ImmutableList.of(BigintType.BIGINT, BooleanType.BOOLEAN), Optional.of(ImmutableList.of("nested_col_0", "nested_col_1")))
+        );
+
+        assertEquals(OrcRecordReader.createStreamReaders(TestCachingOrcDataSource.FakeOrcDataSource.INSTANCE, types, null, includedColumnsGood1).length, 4);
+        assertEquals(OrcRecordReader.createStreamReaders(TestCachingOrcDataSource.FakeOrcDataSource.INSTANCE, types, null, includedColumnsGood2).length, 4);
+        assertThrows(() -> OrcRecordReader.createStreamReaders(TestCachingOrcDataSource.FakeOrcDataSource.INSTANCE, types, null, includedColumnsBad1));
```
Comment (haozhun) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java]: write your own function that `try { ... } catch (PrestoException e) { assertEquals(..., HIVE_BAD_DATA); }`

File: presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                3, new RowType(ImmutableList.of(DoubleType.DOUBLE), Optional.of(ImmutableList.of("nested_col_0")))
+        );
+        final Map<Integer, Type> includedColumnsBad5 = ImmutableMap.of(
+                0, BigintType.BIGINT,
+                1, new MapType(VarcharType.createVarcharType(2), DateType.DATE),
+                2, new ArrayType(DoubleType.DOUBLE),
+                3, new RowType(ImmutableList.of(BigintType.BIGINT, BooleanType.BOOLEAN), Optional.of(ImmutableList.of("nested_col_0", "nested_col_1")))
+        );
+
+        assertEquals(OrcRecordReader.createStreamReaders(TestCachingOrcDataSource.FakeOrcDataSource.INSTANCE, types, null, includedColumnsGood1).length, 4);
```
Comment (haozhun) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java]: It seems unnecessary to assert that size of the return value. Just do `OrcRecordReader.createStreamReaders(TestCachingOrcDataSource.FakeOrcDataSource.INSTANCE, types, null, includedColumnsGood1)`

**Comment by haozhun** [comment]
Please address the comments. Sorry I edited some of my comments. Therefore the one you received in email might not be up to date.

**Comment by geraint0923** [comment]
I tried
`throw new PrestoException(HIVE_BAD_DATA, "...");`
But it seems that it will make presto-orc depends on presto-hive, which will lead to a circular dependancy. Is there any way to use this error code without introducing this issue?

**Comment by geraint0923** [comment]
Use `StandardErrorCode.INVALID_FUNCTION_ARGUMENT` instead of `HIVE_BAD_DATA`. I will update it when I could get `HIVE_BAD_DATA` in without introducing the circular dependancy.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java (hunk 2 above)
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: No, this is the ORC module, not the Hive connector.

**Comment by electrum** [comment]
Use OrcCorruptionException

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 11]
```diff
@@ -393,12 +399,72 @@ private void advanceToNextStripe()
         for (int columnId = 0; columnId < rowType.getFieldCount(); columnId++) {
             if (includedColumns.containsKey(columnId)) {
                 StreamDescriptor streamDescriptor = streamDescriptors.get(columnId);
+
+                Type type = includedColumns.get(columnId);
+                if (!validateOrcType(streamDescriptor, type.getTypeSignature())) {
+                    throw new PrestoException(StandardErrorCode.INVALID_FUNCTION_ARGUMENT, "OrcType and Type do not match");
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Don't use PrestoException in the ORC reader

**Comment by electrum** [comment]
Can you explain the motivation for this commit? I'm not sure this is the right thing to do in this code. The fact that we needed hacks for Raptor (which uses ORC differently the Hive). Is a big red flag. @dain and I need to review this.

**Comment by geraint0923** [comment]
This pull request wants to solve #4695 and it is trying to make sure the types read from orc file match the given types (includedColumns) in `OrcRecordReader.createStreamReaders`.

**Comment by electrum** [comment]
#4695 is this PR?

**Comment by geraint0923** [comment]
I am sorry, just a typo, it is issue #4630.

**Comment by electrum** [comment]
I think it is #4630, through there's no description of the problem.

If I'm understanding correctly, a better way to solve this is to pass in OrcType rather than Presto type and let the users of the reader do the type conversion.

**Comment by haozhun** [comment]
I discussed with @dain on this one. And we agreed that this is a reasonable approach. Admittedly, I didn't realize the issue associated with Raptor at the time.

The motivation of the change is uncategorized errors as a result of corrupt file (file metadata doesn't match partition metadata from Hive metastore)

**Comment by haozhun** [comment]
@geraint0923:

* My Previous comment told you to use PrestoException instead of IAE/ISE. But we should actually use OrcCorruptionException instead. 
* I discussed with @electrum on this pull request. This change you are trying to introduce made us realize we had been doing it wrong. We shouldn't have passed in DATE or TIMESTAMP as the type for raptor readers in the first place.

This bug fix would become a slightly-larger one as a result. But we want to fix it the correct way. Here are the details:
* Find call sites of `OrcReader.createRecordReader` in Raptor. Since we are storing DATE and TIMESTAMP as BIGINT in raptor. We really should have declared to ORC reader that we are looking for a BIGINT instead of TIMESTAMP or BIGINT. Therefore, a new method need to be introduced in Raptor to do the translation recursively. And `OrcReader.createRecordReader` should be invoked with the translated type. This way, we won't need to do the hack inside ORC reader type validation code any more.

**Comment by geraint0923** [comment]
Addressed the comments, added translate method in OrcStorageManager in a separated commit in this pull request and updated the tests.

**Comment by haozhun** [comment]
Move "Translate the type to match OrcType in OrcStorageManager" before "Validate type against StreamDescriptor in OrcRecordReader". When you have multiple commits, you want to make sure every commit is a good one. Without reordering, test fails after your validation commit.

**Comment by haozhun** [comment]
Make sure you address all the comments before you push your new commits. Github will not show the comments associated with commits (as opposed to the pull request) in this pull request when those commits are overwritten.

**Comment by geraint0923** [comment]
Addressed all the comments.

**Comment by haozhun** [comment]
I have a few new comments

**Comment by geraint0923** [comment]
@electrum Is there any other place I need to fix? Thanks!

File: presto-orc/src/test/java/com/facebook/presto/orc/DummyOrcDataSource.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.orc;
+
+import io.airlift.slice.FixedLengthSliceInput;
+
+import java.io.IOException;
+import java.util.Map;
+
+public class DummyOrcDataSource
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/DummyOrcDataSource.java]: Let's keep the name `FakeOrcDataSource` since they both mean the same thing. Then the commit can be titled "Make FakeOrcDataSource a top level class"

File: presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+        assertExceptionInCreateStreamReaders(types, includedColumnsBad5);
+        assertExceptionInCreateStreamReaders(types, includedColumnsBad6);
+        assertExceptionInCreateStreamReaders(types, includedColumnsBad7);
+    }
+
+    public static void assertExceptionInCreateStreamReaders(List<OrcType> types, Map<Integer, Type> includedColumns)
+    {
+        try {
+            OrcRecordReader.createStreamReaders(DummyOrcDataSource.INSTANCE, types, null, includedColumns);
+            fail("PrestoException is expected!");
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java]: Make the message "expected PrestoException"

Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java]: Make the message "expected exception"

File: presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+                new OrcType(OrcType.OrcTypeKind.DOUBLE, ImmutableList.of(), ImmutableList.of()),
+                // 9
+                new OrcType(OrcType.OrcTypeKind.BOOLEAN, ImmutableList.of(), ImmutableList.of())
+        );
+
+        Map<Integer, Type> includedColumnsGood1 = ImmutableMap.of(
+                0, BigintType.BIGINT,
+                1, new MapType(VarcharType.createVarcharType(2), BigintType.BIGINT),
+                2, new ArrayType(DoubleType.DOUBLE)
+        );
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java]: Put the closing `);` on previous line

File: presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.fail;
+
+public class TestOrcRecordReader
+{
+    @Test
+    public void testValidateOrcType()
+            throws OrcCorruptionException
+    {
+        List<OrcType> types = ImmutableList.of(
```
Comment (electrum) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestOrcRecordReader.java]: Use a `.builder()` here with `.add()`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java [hunk 16]
```diff
@@ -188,6 +189,31 @@ public void shutdown()
         deletionExecutor.shutdownNow();
     }
 
+    private static TypeSignature toStorageType(TypeSignature typeSignature)
+    {
+        List<TypeSignatureParameter> typeSignatureParameterList = typeSignature.getParameters();
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java]: Name the local variable `parameters`. No need for such a long variable when there's only one.

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java [hunk 17]
```diff
@@ -188,6 +189,31 @@ public void shutdown()
...
+                checkArgument(typeSignatureParameterList.size() == 1, "ARRAY type must have one element");
+                return new TypeSignature(StandardTypes.ARRAY, ImmutableList.of(TypeSignatureParameter.of(toStorageType(typeSignatureParameterList.get(0).getTypeSignature()))));
+            case StandardTypes.MAP:
+                checkArgument(typeSignatureParameterList.size() == 2, "MAP type must have two elements");
+                return new TypeSignature(StandardTypes.MAP, ImmutableList.of(
+                        TypeSignatureParameter.of(toStorageType(typeSignatureParameterList.get(0).getTypeSignature())),
+                        TypeSignatureParameter.of(toStorageType(typeSignatureParameterList.get(1).getTypeSignature()))));
+            case StandardTypes.ROW:
+                ImmutableList.Builder<TypeSignatureParameter> builder = ImmutableList.builder();
+                for (TypeSignatureParameter typeSignatureParameter : typeSignatureParameterList) {
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java]: This loop variable can be `parameter`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java [hunk 18]
```diff
@@ -188,6 +189,31 @@ public void shutdown()
...
+                return BIGINT.getTypeSignature();
+            case StandardTypes.ARRAY:
+                checkArgument(typeSignatureParameterList.size() == 1, "ARRAY type must have one element");
+                return new TypeSignature(StandardTypes.ARRAY, ImmutableList.of(TypeSignatureParameter.of(toStorageType(typeSignatureParameterList.get(0).getTypeSignature()))));
+            case StandardTypes.MAP:
+                checkArgument(typeSignatureParameterList.size() == 2, "MAP type must have two elements");
+                return new TypeSignature(StandardTypes.MAP, ImmutableList.of(
+                        TypeSignatureParameter.of(toStorageType(typeSignatureParameterList.get(0).getTypeSignature())),
+                        TypeSignatureParameter.of(toStorageType(typeSignatureParameterList.get(1).getTypeSignature()))));
+            case StandardTypes.ROW:
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java]: We can remove this since Raptor doesn't support storing rows

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java [hunk 19]
```diff
@@ -188,6 +189,31 @@ public void shutdown()
         deletionExecutor.shutdownNow();
     }
 
+    private static TypeSignature toStorageType(TypeSignature typeSignature)
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java]: We normally put methods last. Move this to be after the `getType` method.

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java [hunk 20]
```diff
@@ -188,6 +189,31 @@ public void shutdown()
...
+                return new TypeSignature(StandardTypes.ARRAY, ImmutableList.of(TypeSignatureParameter.of(toStorageType(typeSignatureParameterList.get(0).getTypeSignature()))));
+            case StandardTypes.MAP:
+                checkArgument(typeSignatureParameterList.size() == 2, "MAP type must have two elements");
+                return new TypeSignature(StandardTypes.MAP, ImmutableList.of(
+                        TypeSignatureParameter.of(toStorageType(typeSignatureParameterList.get(0).getTypeSignature())),
+                        TypeSignatureParameter.of(toStorageType(typeSignatureParameterList.get(1).getTypeSignature()))));
+            case StandardTypes.ROW:
+                ImmutableList.Builder<TypeSignatureParameter> builder = ImmutableList.builder();
+                for (TypeSignatureParameter typeSignatureParameter : typeSignatureParameterList) {
+                    builder.add(TypeSignatureParameter.of(toStorageType(typeSignatureParameter.getTypeSignatureOrNamedTypeSignature().get())));
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java]: I don't see `getTypeSignatureOrNamedTypeSignature()` used outside of `TypeRegistry`. Why are you using it here?

**Comment by electrum** [comment]
Change commit message: "Move FakeOrcDataSource a top level class" -> "Make FakeOrcDataSource a top level class"

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java [hunk 21]
```diff
@@ -442,6 +444,32 @@ private Type getType(List<OrcType> types, int index)
...
+                return new TypeSignature(StandardTypes.ARRAY, ImmutableList.of(TypeSignatureParameter.of(toStorageType(parameters.get(0).getTypeSignature()))));
+            case StandardTypes.MAP:
+                checkArgument(parameters.size() == 2, "MAP type must have two elements");
+                return new TypeSignature(StandardTypes.MAP, ImmutableList.of(
+                        TypeSignatureParameter.of(toStorageType(parameters.get(0).getTypeSignature())),
+                        TypeSignatureParameter.of(toStorageType(parameters.get(1).getTypeSignature()))));
+            case StandardTypes.ROW:
+                ImmutableList.Builder<TypeSignatureParameter> builder = ImmutableList.builder();
+                for (TypeSignatureParameter parameter : parameters) {
+                    NamedTypeSignature namedTypeSignature = parameter.getNamedTypeSignature();
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java]: Inline this variable

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 22]
```diff
@@ -381,10 +384,12 @@ private void advanceToNextStripe()
         }
     }
 
-    private static StreamReader[] createStreamReaders(OrcDataSource orcDataSource,
+    @VisibleForTesting
+    static StreamReader[] createStreamReaders(OrcDataSource orcDataSource,
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Wrap the first parameter, too

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 23]
```diff
@@ -393,12 +398,70 @@ private void advanceToNextStripe()
...
+                    throw new OrcCorruptionException("ORC type from file footer (%s) does not match expected type of table (%s)", types, includedColumns);
+                }
+
                 streamReaders[columnId] = StreamReaders.createStreamReader(streamDescriptor, hiveStorageTimeZone);
             }
         }
         return streamReaders;
     }
 
+    private static boolean validateOrcType(StreamDescriptor streamDescriptor, TypeSignature type)
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Name this `validOrcType` (short for `isValidOrcType`) because it is returning a boolean. A method named "validate" indicates that it does validation and will throw an exception on failure.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 24]
```diff
@@ -393,12 +398,70 @@ private void advanceToNextStripe()
...
+    {
+        OrcTypeKind orcTypeKind = streamDescriptor.getStreamType();
+        List<StreamDescriptor> streamDescriptorList = streamDescriptor.getNestedStreams();
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Shorten this name to `descriptors`

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java (hunk 23 above)
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Rename the parameter to `typeSignature`

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 25]
```diff
@@ -393,12 +398,70 @@ private void advanceToNextStripe()
...
+
                 streamReaders[columnId] = StreamReaders.createStreamReader(streamDescriptor, hiveStorageTimeZone);
             }
         }
         return streamReaders;
     }
 
+    private static boolean validateOrcType(StreamDescriptor streamDescriptor, TypeSignature type)
+    {
+        OrcTypeKind orcTypeKind = streamDescriptor.getStreamType();
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Name this `type`

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 26]
```diff
@@ -393,12 +398,70 @@ private void advanceToNextStripe()
...
+        List<StreamDescriptor> streamDescriptorList = streamDescriptor.getNestedStreams();
+        List<TypeSignatureParameter> typeSignatureParameterList = type.getParameters();
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Name this `parameters`

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 27]
```diff
@@ -393,12 +398,70 @@ private void advanceToNextStripe()
...
+        switch (type.getBase()) {
+            case StandardTypes.VARCHAR:
+                return (orcTypeKind == OrcTypeKind.STRING || orcTypeKind == OrcTypeKind.VARCHAR || orcTypeKind == OrcTypeKind.CHAR) &&
+                        streamDescriptorList.size() == 0 &&
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Use `isEmpty()` rather than `size() == 0`

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 28]
```diff
@@ -393,12 +398,70 @@ private void advanceToNextStripe()
...
+                return orcTypeKind == OrcTypeKind.MAP &&
+                        streamDescriptorList.size() == 2 &&
+                        typeSignatureParameterList.size() == 2 &&
+                        validateOrcType(streamDescriptorList.get(0), typeSignatureParameterList.get(0).getTypeSignature()) &&
+                        validateOrcType(streamDescriptorList.get(1), typeSignatureParameterList.get(1).getTypeSignature());
+            case StandardTypes.ROW:
+                if (orcTypeKind != OrcTypeKind.STRUCT || streamDescriptorList.size() != typeSignatureParameterList.size()) {
+                    return false;
+                }
+                for (int i = 0; i < streamDescriptorList.size(); ++i) {
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Use `i++`

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 29]
```diff
@@ -393,12 +398,70 @@ private void advanceToNextStripe()
...
+                        return false;
+                    }
+                }
+                return true;
+            case StandardTypes.BOOLEAN:
+                return orcTypeKind == OrcTypeKind.BOOLEAN;
+            case StandardTypes.DOUBLE:
+                return orcTypeKind == OrcTypeKind.FLOAT ||
+                        orcTypeKind == OrcTypeKind.DOUBLE;
+            case StandardTypes.BIGINT:
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: This might need to be updated for the newly added integer types. cc @raghavsethi

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 30]
```diff
@@ -393,12 +398,70 @@ private void advanceToNextStripe()
...
+                return orcTypeKind == OrcTypeKind.BYTE ||
+                        orcTypeKind == OrcTypeKind.SHORT ||
+                        orcTypeKind == OrcTypeKind.INT ||
+                        orcTypeKind == OrcTypeKind.LONG ||
+                        orcTypeKind == OrcTypeKind.DECIMAL;
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: I'm not sure this is correct, especially after the decimal changes. cc @martint

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 31]
```diff
@@ -393,12 +398,70 @@ private void advanceToNextStripe()
...
+            case StandardTypes.DATE:
+                return orcTypeKind == OrcTypeKind.DATE;
+            case StandardTypes.TIMESTAMP:
+                return orcTypeKind == OrcTypeKind.TIMESTAMP;
+            case StandardTypes.VARBINARY:
+                return orcTypeKind == OrcTypeKind.BINARY;
```
Comment (electrum) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Should this have the same validation is `VARCHAR` for `descriptors.isEmpty()`?
